ons-spark/spark-analysis/Cramer_v.md

```diff
@@ -0,0 +1,439 @@
+$ Hedgehog                                           <dbl> 0, 0, 0, 0, 0, 0, 0…
+$ Horse                                              <dbl> 12, 16, 12, 12, 22,…
+$ Lamb                                               <dbl> 0, 0, 0, 0, 0, 1, 0…
+$ Lizard                                             <dbl> 0, 0, 1, 0, 0, 0, 0…
+$ Pigeon                                             <dbl> 1, 1, 1, 0, 0, 0, 0…
+$ Rabbit                                             <dbl> 3, 0, 2, 0, 0, 0, 0…
+$ Sheep                                              <dbl> 0, 0, 0, 1, 0, 1, 0…
+$ Snake                                              <dbl> 0, 2, 1, 0, 0, 0, 0…
+$ Squirrel                                           <dbl> 2, 7, 5, 5, 4, 4, 1…
+$ Tortoise                                           <dbl> 0, 0, 0, 1, 0, 0, 0…
+$ `Unknown - Animal Rescue From Water - Farm Animal` <dbl> 1, 0, 0, 0, 0, 1, 1…
+$ `Unknown - Domestic Animal Or Pet`                 <dbl> 29, 19, 7, 29, 13, …
+$ `Unknown - Heavy Livestock Animal`                 <dbl> 0, 2, 2, 0, 8, 4, 0…
+$ `Unknown - Wild Animal`                            <dbl> 7, 12, 5, 7, 0, 4, …
+```
+````
+Now that we have converted out data into a contingency table, we need to be careful about converting this into a matrix or array type variable. If we do this without considering the `CalYear_animal_type` column, we will end up with some string within our matrix. We would spot this issue when moving forward to use `.chi2_contingency()`, as we would get a `TypeError` raised. We need to find some way of dealing with our column, such that it is not converted when changing our pandas DataFrame into a numpy array. We do this by setting the `cal_year_animal_type` column as the index of our pandas dataframe, as neither the index or column headers are converted into a numpy array, just the internal values are extracted.
+
+A type issue is also raised in R, however we will deal with this slightly differently. Instead of changing the index of our DataFrame, we will simply drop the `cal_year_animal_type` column. This allows us to pass the frequency table directly into the `chisq.test()` function.
+````{tabs}
+```{code-tab} py
+freq_pandas_reindexed = freq_pandas.set_index('cal_year_animal_type')
+freq_numpy = np.array(freq_pandas_reindexed)
+
+```
+
+```{code-tab} r R
+
+freq_r <- freq_spark %>% collect() 
+freq_r <- subset(freq_r, select = -cal_year_animal_type)
+
+
+```
+````
+As we are wanting to calculate Cramér's V for two examples, we will define a function to perform the $\chi^2$ test and calculate the statistic. We only need the $\chi^2$ statistic, as such we extract this from the `chi2_contingency`/`chisq.test` functions using `[0]` in python and `$statistic` in R. `chisq.test` has additional text included in the output, to remove this we cast this output as a double, keeping only the numerical value.
+````{tabs}
+```{code-tab} py
+def get_cramer_v(freq_numpy):
```

Comment:
Docstrings would be good, also what is the expected input ? It looks like a numpy array , does it work with other inputs ? like a table or a dataframe?  Also what returns , a double ?